File: src/cspom/extension/ExtensionConstraint.java
package cspom.extension;

import java.util.List;
import java.util.Map;

import cspom.constraint.AbstractConstraint;
import cspom.variable.Variable;

public class ExtensionConstraint extends AbstractConstraint {

	private final Extension relation;

	public ExtensionConstraint(String name, List<Variable> scope,
			Extension relation) {
		super(name, scope);
		this.relation = relation;
	}

	public Extension getRelation() {
		return relation;
	}

	public String toString() {
		return super.toString() + ": " + relation;
	}

	public boolean evaluate(final Number[] numbers) {
		return relation.evaluate(numbers);
	}

	public ExtensionConstraint standardize(final List<Variable> scope) {
		final int[] newPosition = new int[getArity()];
		final Map<Variable, Variable> newOrder = newOrder(scope);

		for (int i = getArity(); --i >= 0;) {
			newPosition[i] = position(newOrder.get(getScope().get(i)));
		}

		return new ExtensionConstraint(getName(), scope, relation
				.reverse(newPosition));
	}

}

File: src/main/scala/cspom/extension/ExtensionConstraint.scala
package cspom.extension
import cspom.constraint.{ PermutableConstraint, CSPOMConstraint }
import cspom.variable.CSPOMVariable
import cspom.xcsp.Extension

/**
 * A class that implements extension constraints, that it constraints defined by
 * a set of allowed or forbidden tuples.
 *
 * @author vion
 *
 * @param <T>
 */
final class ExtensionConstraint(
  val relation: HashTrie,
  val init: Boolean,
  scope: Seq[CSPOMVariable])
  extends CSPOMConstraint("ext", scope) with PermutableConstraint {

  def this(relation: HashTrie, init: Boolean, scope: Array[CSPOMVariable]) =
    this(relation, init, scope.toSeq)

  override def toString = "%s: %s (%s)".format(super.toString, relation, if (init) "forbidden" else "allowed");

  override def evaluate(tuple: Seq[_]) = init ^ relation.contains(tuple.map(_.asInstanceOf[Int]).toArray)

  //  override lazy val hashCode = 31 * super.hashCode + relation.hashCode
  //
  //  override def equals(obj: Any) = obj match {
  //    case ec: ExtensionConstraint => super.equals(ec) && ec.relation == this.relation
  //    case _ => false
  //  }

  def standardize(newScope: Seq[CSPOMVariable]) = {
    assert(newScope.size == arity)
    new ExtensionConstraint(relation.permute(scope.map(v => newScope.indexOf(v))), init, newScope.toList)
  }

  override def replacedVar(which: CSPOMVariable, by: CSPOMVariable) = {
    new ExtensionConstraint(relation, init, scope map { v => if (v == which) by else v })
  }

}